### ISSUE TITLE
[Question] Is there a clean way of swapping pages in pikepdf?

### ISSUE THREAD

**Issue by m-holger** [pikepdf/pikepdf#192]
I am looking for a way to swap pages to give the following result

``` python
print([p.objgen for p in pdf.pages])
[(1, 0), (4, 0), (7, 0), (10, 0)]  # before
[(1, 0), (7, 0), (4, 0), (10, 0)] # after
```
The only way I can think of is something like

``` python
def index(array, obj):
    i = 0
    while array[i] != obj:
        i +=1
    return i

def swap(pdf, page1, page2):
    p1 = page1.Parent
    i1 = index(p1.Kids, page1)
    p2 = page2.Parent
    i2 = index(p2.Kids, page2)
    page1.Parent = p2
    p2.Kids[i2] = page1
    page2.Parent = p1
    p1.Kids[i1] = page2  
```
This seems to work, but leaves pdf.pages inconsistent with the page tree

``` python
print([p.objgen for p in pdf.pages])
[(1, 0), (4, 0), (7, 0), (10, 0)] 
print([p.objgen for p in pdf.Root.Pages.Kids])
[(1, 0), (7, 0), (4, 0), (10, 0)]
```

Is there a better way of swapping pages in pikepdf?

**Comment by jbarlow83** [comment]
`pdf.pages[1:3] = pdf.pages[3:1:-1]` should do it.

Or use tuple swapping: x, y = y, x

**Comment by m-holger** [comment]
Thanks for the speedy reply.

With slices I get
``` python
pdf.pages[1:3] = pdf.pages[3:1:-1]
[(1, 0), (4, 0), (7, 0), (10, 0)]  # before
[(1, 0), (24, 0), (25, 0), (10, 0)] # after
```
i.e not a swap but a replacement of pages 2 & 3 with copies of pages 3 & 2 (thus invalidating internal links). 
I am trying to move pages without invalidating any Names pointing (in this case ) to (4, 0) and (7, 0). Intuitively I feel that this should be simple (perhaps requiring a replace counterpart to emplace, i.e. maintaining the objgen of the source rather than the destination) but somehow I cannot find  a clean way of doing this with pikepdf.

**Comment by jbarlow83** [comment]
Oh, I see. Try `Pdf._swap_objects()`.

**Comment by m-holger** [comment]
@jbarlow83 Sorry, did not work. 

It successfully swapped the pages, but not the objgen, i.e. the content was reordered from [A, B, C ,D] to [A, C, B, D], but objgens stayed constant as [(1, 0), (4, 0), (7, 0), (10, 0)]

Clicking onto a link to page C now lands on page B.

However, your suggestion inspired the following solution
``` python
def swap(pdf, page1, page2):
    p1 = pdf.pages[page1]
    pdf.pages[page1] = p1
    p2 = pdf.pages[page2]
    pdf.pages[page2] = p2
    pdf._add_page_at(p1, True, pdf.pages[page2])
    del(pdf.pages[page1])
    pdf._add_page_at(p2, True, pdf.pages[page1])
    del(pdf.pages[page2])
```
Do you think it might be useful to add move, swap and replace methods that do not break existing links to pikepdf ?

Btw, I spotted that emplace (incorrectly) replaces the /Parent with the /Parent of the source. Do you want me to open an issue, submit a pull request or leave it to you?

**Comment by jbarlow83** [comment]
I fixed the /Parent - emplace issue - good catch - that's quite subtle given pikepdf's flattening of the page tree by default.

The author of QPDF is planning to overhaul the page management interface so I am going to table any rework on this until that is available. 
See: https://github.com/qpdf/qpdf/issues/514#issuecomment-803195856

The existing API does not support the operations you mention - it only has insert and delete, and it's possible for insert to construct a copy of the provided object and insert that rather than the object provided.

I suppose the best way to do this in pikepdf currently, without using private APIs, would be to swap the elements of page1 and page2 except Parent similar to what `emplace` does. This would look like:
- make shallow copy_of_page1
- page1.emplace(page2)
- page2.emplace(copy_of_page1)
- delete copy_of_page1

**Comment by m-holger** [comment]
@jbarlow83 Sorry, but this does not work 

If you start of with

```
Name Page_A (4,0)
Name Page_B (5,0)
Page 1: (4,0) -> Content of Page A with link to Page_B (5,0), i.e. Page B
Page 2: (5,0) -> Content of Page B with link to Page_A (4,0), i.e. Page A
```

you end up with

```
Name Page_A (4,0)
Name Page_B (5,0)
Page 1: (4,0) -> Content of Page B with link to Page_A (4,0), i.e. Page B
Page 2: (x,0) -> Content of Page A with link to Page_B (5,0), i.e. to undefined
where x not in {4,5}
```

What we need is 

```
Name Page_A (4,0)
Name Page_B (5,0)
Page 1: (5,0) -> Content of Page B with link to Page_A (4,0), i.e. Page A
Page 2: (4,0) -> Content of Page A with link to Page_B (5,0), i.e. Page B
```



I think the problem lies with PageList::insert_page


``` cpp
void PageList::insert_page(size_t index, QPDFObjectHandle h)
{
    // Find out who owns us
    QPDF *handle_owner = h.getOwningQPDF();
    QPDFObjectHandle page;
    bool copied = false;

    if (handle_owner == this->qpdf.get() || !handle_owner) {
        // qpdf does not accept duplicating pages within the same file,
        // so manually create a copy
        page = this->qpdf->makeIndirectObject(h.shallowCopy());
        copied = true;
    } else {
        page = h;
    }
```

Consider 

``` python
print([p.objgen for p in pdf.pages])

p1, p2 = pdf.pages[1], pdf.pages[2] 

# remove p1 & p2 from page tree
pdf.pages[1] = pdf.pages[3]
pdf.pages[2] = pdf.pages[3]

print([p.objgen for p in pdf.pages])
print(p1.is_owned_by(pdf))

pdf.pages[2], pdf.pages[1] = p1, p2

print([p.objgen for p in pdf.pages])



```

with output

```
[(1, 0), (4, 0), (7, 0), (10, 0)]
[(1, 0), (24, 0), (25, 0), (10, 0)]
True
[(1, 0), (27, 0), (26, 0), (10, 0)]

```

Here (4,0) and (7, 0) are removed from the page tree prior to reinsertion
and therefore they can be safely reinserted. However, give that they 
are owned by the pdf, they are replaced by copies anyway - breaking existing 
links.

Given that qpdf >= 10.3.1 now checks for duplicate pages and creates copies 
when necessary, I think an unconditional `page = h` would solve the problem.

**Comment by jbarlow83** [comment]
The improvements to duplicate page checking are not available in any released version of qpdf yet so I cannot release a change against that.

This method seems to work currently, unless I'm missing something:
```python
p1, p2 = pdf.pages[1], pdf.pages[2]
print([p.objgen for p in pdf.pages])
pdf.pages.append(p1)
p1.emplace(p2)
p2.emplace(pdf.pages[-1])
del pdf.pages[-1]
print([p.objgen for p in pdf.pages])
pdf.save(...)
```

**Comment by m-holger** [comment]
I ran your code on [in.pdf](https://github.com/pikepdf/pikepdf/files/6295374/in.pdf) to produce [out.pdf](https://github.com/pikepdf/pikepdf/files/6295377/out.pdf)

To illustrate the problem, go to index and click on page 64 ( `__new__()` )



> The improvements to duplicate page checking are not available in any released version of qpdf

Yes, sorry, found that out the hard way. At the moment we still need to chack that the inserted page is not in pdf.pages. The check that pdf.pages owns the inserted page goes too far.

**Comment by m-holger** [comment]
@jbarlow83 

Given that qpdf 10.4.0 checks for duplicate pages,  would it make sense to implement the change suggested in https://github.com/pikepdf/pikepdf/issues/192#issuecomment-817297764

**Comment by jbarlow83** [comment]
Probably, but pikepdf still supports qpdf 10.3.1 and the fix needs 10.4.0. Maybe warnings will suffice for the older version.

**Comment by jbarlow83** [comment]
Tried with 10.4.0 and unconditional `page = h`. This still doesn't have the desired effect. QPDF prefers to copy when inserting an existing page.

I did find a working approach - it requires direct manipulation of the /Pages tree to change the order without breaking other references. I'm not sure it will be possible to get this right in idiomatic Python, i.e. some special swap function is probably needed, because statements like `a, b = b, a` are not transactional - an extension library cannot "see" that Python is performing a swap.

**Comment by m-holger** [comment]
In my experience direct manipulation of the page tree is not necessary (& imo undesirable).

The key requirement is that after you extract the page it must be deleted from the page tree before re-insertion into the page tree. Happy to help if I can be of any assistance.

**Comment by m-holger** [comment]
That said, I am not sure that `a, b = b, a` can be easily fixed. The best solution for that one is probably a section in the manual on the does and don'ts if you don't want to mess up internal links. (Again, if this is something you want to do I'd be happy to assist).

I am not sure about a `swap` method, as that deals only with one special case. A pair of `extract` and `reinsert` methods would be more flexible.

**Comment by m-holger** [comment]
The other method that would be useful would be methods to copy all pages to a temporary list and to empty the page tree (without deleting the `pages` dictionary itself), which would deal with applications who assemble effectively a new file from pages of the existing (plus additional) files. This is what, at least conceptually, both qpdf and PdfArranger do.

**Comment by ffeldhaus** [comment]
@jbarlow83 Could you summarize the recommended way to swap pages? Is it by using the private function `PDF._swap_objects` or is there a better way? Why is the function private? Is there any outstanding work to support swapping pages?

**Comment by jbarlow83** [comment]
This is now supported: reorder pages by removing and reinserting them (`del pdf.pages[i]` then `pdf.pages.insert(j, page)`), which preserves each page's objgen so internal references (named destinations, links, outlines) survive. Documented at https://pikepdf.readthedocs.io/en/latest/topics/pages.html#reordering-pages-without-breaking-links
